### DISCUSSION TITLE
Most needed/wanted features

### DISCUSSION THREAD

**Discussion by asmyshlyaev177** [asmyshlyaev177/react-horizontal-scrolling-menu#221] [Polls]
Wanna know what most demanded feature of this library and how can improve it.

Can write your own option in comments.

**Comment by juanlatorre** [comment]
Maybe mention how to make it work with virtualization?

**Comment by keegan-black** [comment]
I would love if column gaps were exposed as a property (if it isn't already)

**Reply by asmyshlyaev177** [reply]
You can just add margin to Item with CSS. Can try it on https://codesandbox.io/p/sandbox/basic-example-swg0y?file=%2Fsrc%2Fcard.tsx%3A15%2C7-15%2C26

**Comment by gavrilikhin-d** [comment]
I would love if it wasn't politically biased

**Comment by noJokesCoder** [comment]
I am trying to implement a badge showing the count of items left to scroll (the ones that are on the right from the last currently visible item) & somehow it is impossible to retrieve this data on mount. Meaning I can get it with onScroll, but oce page loaded - this data is not available anywhere

**Reply by btnalexandre** [reply]
same problem for me the data is not available on mount

**Reply by asmyshlyaev177** [reply]
Can use `apiRef` ref. https://codesandbox.io/p/sandbox/apiref-vdr0d?file=%2Fsrc%2Findex.tsx

**Reply by btnalexandre** [reply]
Thanks for your quick reply. I used `apiRef`. When i want to get all items ids with `apiRef.current.items.toItems()` it's return `undefined` after the load of the page. 

When I made a hmr update with a little change in the code, the function toItems() seems to return all items that I want.

So I suspect a bug with the state

**Reply by asmyshlyaev177** [reply]
Sure, something should work.
Related example https://asmyshlyaev177.github.io/react-horizontal-scrolling-menu/?path=/story/examples-progress--progress

**Reply by btnalexandre** [reply]
Thanks !

**Comment by edgaraparicioglobant** [comment]


**Reply by asmyshlyaev177** [reply]
Because `apiRef` is ref, it's always same instance, useEffect will run only once. Try to use it in a callback.

**Reply by eBaeza** [reply]
sorry i delete the comment, i use the incorrect account hehe, but i'm just looking a way to set a default to scroll into on init. Using the callback the items always changes, doing something like this. If i use the onInit event the items aren't ready yet.

```
const handleRef = useCallback(
    (apiRef: publicApiType) => {
      console.log(apiRef.items.toItems());
      if (apiRef.items.size > 0) {
        const index = items.findIndex((item) => pathname.includes(item.path));
        if (index !== -1) {
          const item = apiRef.getItemElementByIndex(String(index));
          // if (item) apiRef.scrollToItem(item, 'auto', 'start');
        }
      }
    },
    [items, pathname]
  );
  ```

**Reply by asmyshlyaev177** [reply]
Something like this should work
```typescript
  const apiRef = React.useRef<publicApiType>();
  React.useEffect(() => {
    const items = apiRef.current?.items.toItems()
    // or .getItemByIndex
    apiRef.current.scrollToItem(apiRef.current.getItemById(items[5]), 'auto')
  }, [apiRef])
```

**Reply by eBaeza** [reply]
don't works for me u.u the items are always empty u.u

**Reply by asmyshlyaev177** [reply]
@eBaeza will take a closer look when I have time.

**Reply by eBaeza** [reply]
thanks for the quickly response, i appreciate. just for the record i'm using the version 8.2.0

**Reply by asmyshlyaev177** [reply]
@eBaeza https://asmyshlyaev177.github.io/react-horizontal-scrolling-menu/?path=/story/examples-scrolltoitem--scroll-to-item , for now pretty ugly, but it require more time.

**Reply by eBaeza** [reply]
thanks, i take a look
